nucypher/blockchain/eth/signers/base.py

```diff
@@ -0,0 +1,94 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+from abc import ABC, abstractmethod
+from typing import List
+from urllib.parse import urlparse
+
+from hexbytes.main import HexBytes
+
+from nucypher.utilities.logging import Logger
+
+
+class Signer(ABC):
+
+    URI_SCHEME = NotImplemented
+    SIGNERS = NotImplemented  # set dynamically in __init__.py
+
+    log = Logger(__qualname__)
+
+    class SignerError(Exception):
+        """Base exception class for signer errors"""
+
+    class InvalidSignerURI(SignerError):
+        """Raised when an invalid signer URI is detected"""
+
+    class AccountLocked(SignerError):
+        def __init__(self, account: str):
+            self.message = f'{account} is locked.'
+            super().__init__(self.message)
+
+    class UnknownAccount(SignerError):
+        def __init__(self, account: str):
+            self.message = f'Unknown account {account}.'
+            super().__init__(self.message)
+
+    @classmethod
+    def from_signer_uri(cls, uri: str) -> 'Signer':
+        parsed = urlparse(uri)
+        scheme = parsed.scheme if parsed.scheme else parsed.path
+        try:
+            signer_class = cls.SIGNERS[scheme]
+        except KeyError:
+            # This block can be considered the "pass-through"
+            # for providers to be used as external signers.
+            try:
+                from nucypher.blockchain.eth.signers.software import Web3Signer
+                signer = Web3Signer.from_signer_uri(uri=uri)
+            except cls.InvalidSignerURI:
+                message = f'{uri} is not a valid signer URI.  Available schemes: {", ".join(cls.SIGNERS)}'
+                raise cls.InvalidSignerURI(message)
```

Comment:
This is a huge anti-pattern, imho.
Implicitly falling back to a Web3 signer seems like something that would occur _outside_ this class and be done explicitly instead.
I'll point out as well that the classes that implement this ABC also do checking to ensure that the URI is valid for the signer (e.g. `TrezorSigner`).

IMHO, when it comes to signing APIs, everything should be explicit and not make assumptions on behavior like this.

Reply:
It's not abundantly obvious - but this is preserving the pre-exsting behavior wherein `--provider` can be used as a signer.  

I don't really characterize this situation as a *fallback* since `signer` can simply be `None` - rather it's a scheme detection *pass-through* -- since there are many ways to specify node comm ( `ipc://`, `file://`, `http://`... see https://github.com/nucypher/nucypher/blob/main/nucypher/blockchain/eth/interfaces.py#L396 ) using the passthrough here allows a simple and DRY usage of existing scheme detection and web3 connection logic.